add headers: compare, cfloat, cinttypes, climits, cstdarg, modify headers: utility, algorithm, new, memory, cstddef

ilobilo opened the pull request:
added headers: cfloat, cinttypes, climits, cstdarg, compare

* compare
  - partial_ordering
  - weak_ordering
  - strong_ordering
  - common_comparison_category/_t
  - compare_three_way_result/_t
  - compare_three_way
  - is_eq
  - is_neq
  - is_lt
  - is_lteq
  - is_gt
  - is_gteq

* cstddef
  - std::ptrdiff_t

* cinttypes
  - std::imaxdiv_t

<!---
* memory
  - std::allocator
-->
* new
  - align_val_t
  - destroying_delete/_t

* algorithm
  - constexpr min/max

* utility
  - piecewise_construct/_t
  - in_place/_t
  - in_place_type/_t
  - in_place_index/_t

ArsenArsen commented on stage2/include/memory:
```diff
@@ -14,11 +16,37 @@ T *addressof(T &r) noexcept {
+		return true;
+	}
+};
+
```
This is not even in P1642, are you sure you want to add it?

ArsenArsen commented on stage2/include/climits:
```diff
@@ -0,0 +1,6 @@
+#define _CXXSHIM_CLIMITS
+
+#include <limits.h>
+
```
these appear to miss `std` aliases, might want to add them (similarly in other headers)

ilobilo replied:
climits and cfloat only have macros in them
cinttypes has std::imaxdiv_t which I added

ArsenArsen commented on stage2/include/cstddef:
```diff
@@ -8,7 +8,8 @@ namespace std {
 enum class [[gnu::may_alias]] byte : unsigned char { };
 
 using nullptr_t = decltype(nullptr);
-using size_t = ::size_t;
```
unrelated change

ArsenArsen commented on stage2/include/memory:
```diff
@@ -14,11 +16,37 @@ T *addressof(T &r) noexcept {
 #else
-	return reinterpret_cast<T *>(
-      const_cast<char *>(&reinterpret_cast<const volatile char &>(r)));
+	return reinterpret_cast<T *>(const_cast<char *>(&reinterpret_cast<const volatile char &>(r)));
```
unrelated change

ArsenArsen commented on stage2/include/utility:
```diff
@@ -63,6 +63,34 @@ using make_index_sequence = make_integer_sequence<size_t, N>;
+
+template<size_t I>
+inline constexpr in_place_index_t<I> in_place_index { };
+
```
this commit summary line is too long (d036ab00fefd31bc1d18f92c3fd8a2b78f9aea3e)

ArsenArsen reviewed (commented):
where does this code come from? the code style seems all over the place

EDIT: unsure how much we want to grow cxxshim anyway, especially outside of the bounds of P1642, considering it'd be effectively obsoleted by GCC13

ArsenArsen commented on stage2/include/compare:
```diff
@@ -0,0 +1,344 @@
+#ifndef _CXXSHIM_COMPARE
```
also this commit message (dda47138d6504c6a471c53621e44005b285c98fc) has an excessively long line

Commit:
utility: modify header
this commit adds: piecewise_construct/_t, in_place/_t,
in_place_type/_t, in_place_index/_t

Commit:
algorithm: make min and max constexpr

Commit:
add headers: cfloat, cinttypes, climits, cstdarg

Commit:
new: add align_val_t, destroying_delete/_t

Commit:
cstddef: add std::ptrdiff_t

Commit:
add header: compare
this commit adds: partial_ordering, weak_ordering, strong_ordering,
common_comparison_category/_t, compare_three_way_result/_t,
compare_three_way, is_eq, is_neq, is_lt, is_lteq, is_gt, is_gteq
